pretty_midi/pretty_midi.py

```diff
@@ -1019,6 +1019,24 @@ def adjust_times(self, original_times, new_times):
         # Get original downbeat locations (we will use them to determine where
         # to put the first time signature change)
         original_downbeats = self.get_downbeats()
+        
+        # Test original_times and new_times for monotonicity
+        # if they are not monotonic, a warning should be given 
+        # and they should be made monotonic with np.maximum.accumulate
+        def _strictly_increasing(L):
+            if all(x<y for x, y in zip(L, L[1:])):
```

Comment:
You're right, using `diff` is twice as fast! I'll replace `zip` with the `diff` approach.

Reply:
So I believe `original_times` must be strictly increasing (i.e. no repeats) - it doesn't make sense to map one time to multiple different times. I think it's ok if `new_times` is monotonic but not strictly increasing, because you can map a span of the original sequence to a single point in the new sequence (basically cropping out part of the sequence).

Based on this I think the correct pattern would be
```Python
original_times, unique_idx = np.unique(original_times, return_index=True)
if (unique_idx.size != original_times.size) or unique_idx != np.arange(unique_idx.size):
    warnings.warn('original_times must be strictly increasing; automatically enforcing this.')
new_times = new_times[unique_idx]
if not np.all(np.diff(new_times) > 0):
    warnings.warn('new_times must be monotonic, automatically enforcing this.')
new_times = np.maximum.accumulate(new_times)
```
Does this make sense? Does it seem right to you?

Reply:
Its probably just me, but I don't quite follow. How is original_times being checked for increasingness? Also, I can't think of a case where `unique_idx.size != original_times.size` -- could you clarify what its checking?

Reply:
Okay that makes sense to me.

I am getting an error with `new_times = new_times[unique_idx]`: `TypeError: only integer scalar arrays can be converted to a scalar index`

I can avoid this error by writing `new_times = np.array(new_times)[unique_idx].

Reply:
> I am getting an error with new_times = new_times[unique_idx]: TypeError: only integer scalar arrays can be converted to a scalar index

Hmm, this should not be happening I don't think. What is the input you are using for `new_times`?

Reply:
I should clarify that the errors shows up during the travis build. The travis build fails for `update new_times if original_times fails check` with the following error during the call to test_adjust_times:

```
ERROR: test_pretty_midi.test_adjust_times
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.5.5/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/craffel/pretty-midi/tests/test_pretty_midi.py", line 165, in test_adjust_times
    pm.adjust_times([0, 5, 5, 10], [5, 10, 12, 17])
  File "/home/travis/build/craffel/pretty-midi/pretty_midi/pretty_midi.py", line 1035, in adjust_times
    new_times = new_times[unique_idx]
TypeError: only integer scalar arrays can be converted to a scalar index
```

Reply:
Yes, it must be because `new_times` is a list.  Can you do
`new_times = np.asarray(new_times)[unique_idx]`
instead of
`new_times = np.array(new_times)[unique_idx]`
`asarray` does not copy the array by default.

Reply:
Okay I have that change in the latest commit. Now I get an error in a different part of the call to test_adjust_times:

   FAIL: test_pretty_midi.test_adjust_times
   ----------------------------------------------------------------------
   Traceback (most recent call last):
     File "/home/travis/virtualenv/python3.5.5/lib/python3.5/site-packages/nose/case.py", line 198, 
   in runTest
       self.test(*self.arg)
     File "/home/travis/build/craffel/pretty-midi/tests/test_pretty_midi.py", line 169, in 
   test_adjust_times
       [n.start for n in pm.instruments[0].notes], expected_starts)
   AssertionError

Running this test locally does not reproduce the error for me, so I'm not sure how to proceed.

Reply:
This makes sense. That test takes a `pretty_midi.PrettyMIDI` object which has notes at start times `[1, 2, 3, ..., 9]` and remaps them as `pm.adjust_times([0, 5, 5, 10], [5, 10, 12, 17])`.  In this new format `np.unique([0, 5, 5, 10])` returns indices `[0, 1, 3]` which correspond to new times`[5, 10, 17]`. The previous behavior of the code mapped 5 to 12, now it maps 5 to 10. Neither feels obviously correct to me so as long as it's documented I think it's fine. You should be able to fix the test by changing https://github.com/craffel/pretty-midi/blob/master/tests/test_pretty_midi.py#L165 to `pm.adjust_times([0, 5, 5, 10], [5, 12, 13, 17])`. Does that work?

Reply:
With `pm.adjust_times([0, 5, 5, 10], [5, 12, 13, 17])` I get

[n.start for n in pm.instruments[0].notes]
Out[194]: [6.4, 7.8, 9.2, 10.6, 12.0, 13.0, 14.0, 15.0, 16.0]

which does not match the test's expectation: `expected_starts = [6, 7, 8, 9, 12, 13, 14, 15, 16]`

Reply:
You're right. One way to do it would be use `[7, 12, 13, 17]` for `new_times` and shift `expected_starts` by 2, I think. Feel free to edit the test to whatever you think makes sense in this case.

Reply:
Okay now I'm passing those tests. 

I'm failing the tests in the complicated example. I'm having trouble following your calculations for the expected starts in that section. It looks like there is some interaction between the linear interpolation from adjust_times and the addition of new time signatures. It will take some time for me to wrap my head around it.